sound/soc/sof/intel/hda-dai.c

```diff
@@ -322,9 +337,20 @@ static int __maybe_unused hda_dai_trigger(struct snd_pcm_substream *substream, i
 static int hda_dai_prepare(struct snd_pcm_substream *substream, struct snd_soc_dai *dai)
 {
 	struct snd_soc_pcm_runtime *rtd = snd_soc_substream_to_rtd(substream);
+	const struct hda_dai_widget_dma_ops *ops = hda_dai_get_ops(substream, dai);
+	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(dai, substream->stream);
 	int stream = substream->stream;
+	int ret;
+
+	ret = hda_dai_hw_params(substream, &rtd->dpcm[stream].hw_params, dai);
```

Comment:
Is there any specific reason that we always call hw_params in .prepare?

Reply:
because sometimes we dont have a hw_free like in the case of xruns and we'll need to do everything again.

Reply:
prepare() should be fast. We should call this out in comments - we need to do X,Y out of cycle because Z